db/db_with_timestamp_basic_test.cc

```diff
@@ -704,6 +704,60 @@ TEST_F(DBBasicTestWithTimestamp, SimpleIterate) {
   Close();
 }
 
+TEST_F(DBBasicTestWithTimestamp, TrimHistoryTest) {
+  Options options = CurrentOptions();
+  options.env = env_;
+  options.create_if_missing = true;
+  const size_t kTimestampSize = Timestamp(0, 0).size();
+  TestComparator test_cmp(kTimestampSize);
+  options.comparator = &test_cmp;
+  DestroyAndReopen(options);
+  // Construct data of different versions with different ts
+  WriteOptions wopts;
+  std::string ts_str = Timestamp(2, 0);
+  Slice ts = ts_str;
+  ASSERT_OK(db_->Put(wopts, "k1", ts, "v1"));
+  ts_str = Timestamp(4, 0);
+  ts = ts_str;
+  ASSERT_OK(db_->Put(wopts, "k1", ts, "v2"));
+  ts_str = Timestamp(5, 0);
+  ts = ts_str;
+  ASSERT_OK(db_->Delete(wopts, "k1", ts));
+  ts_str = Timestamp(6, 0);
+  ts = ts_str;
+  ASSERT_OK(db_->Put(wopts, "k1", ts, "v3"));
+  ASSERT_OK(Flush());
+  Close();
+
+  ColumnFamilyOptions cf_options(options);
+  std::vector<ColumnFamilyDescriptor> column_families;
+  column_families.push_back(
+      ColumnFamilyDescriptor(kDefaultColumnFamilyName, cf_options));
+  DBOptions db_options(options);
+  // Trim data whose version is newer than Timestamp(3, 0), which means
+  // only (k1, v1) should survive after call DB::OpenAndTrimHistory.
+  ts_str = Timestamp(3, 0);
+  ASSERT_OK(DB::OpenAndTrimHistory(db_options, dbname_, column_families,
+                                   &handles_, &db_, ts_str));
+  ReadOptions ropts;
+  ts_str = Timestamp(2, 0);
+  ts = ts_str;
+  ropts.timestamp = &ts;
+  std::string value;
+  Status s = db_->Get(ropts, "k1", &value);
+  ASSERT_OK(s);
+  ASSERT_EQ("v1", value);
+
+  ts_str = Timestamp(7, 0);
+  ts = ts_str;
+  ropts.timestamp = &ts;
+  s = db_->Get(ropts, "k1", &value);
+  ASSERT_OK(s);
+  // read with ts=7 should only see (k1, v1)
+  ASSERT_EQ("v1", value);
+  Close();
```

Comment:
I think another slightly more interesting case is as follows.
```
// t1 < t2
put(k, t1, v1), del(k, t2)
// read(k,t2) <-- NOT_FOUND
trim(t2)
// read(k,t2) <-- found <k, t1, v1>
```

Reply:
Add this case,  when OpenAndTrimHistory with trim_ts (4,0), the k1 is found